File: employee.cpp
#include "employee.hpp"
#include <iostream>

Employee::Employee()
	: m_fullName("")
	, m_age(0)
	, m_contactNumber("")
	, m_address("")
	, m_position("")
	, m_hoursWorked(0.0)
	, m_hourlyRate(0.0)
	, m_login("")
	, m_password("")
	, m_salaryBalance(0.0)
	, m_salesCount(0)
{
}

Employee::Employee(const std::string& fullName, int age, const std::string& contactNumber,
				   const std::string& address, const std::string& position,
				   double hourlyRate, const std::string& login, const std::string& password)
	: m_fullName(fullName)
	, m_age(age)
	, m_contactNumber(contactNumber)
	, m_address(address)
	, m_position(position)
	, m_hoursWorked(0.0)
	, m_hourlyRate(hourlyRate)
	, m_login(login)
	, m_password(password)
	, m_salaryBalance(0.0)
	, m_salesCount(0)
{
}

double Employee::CalculateSalary() const
{
	// Расчет з/п = отработанные часы * почасовая ставка + баланс (премии/штрафы)
	return m_hoursWorked * m_hourlyRate + m_salaryBalance;
}

void Employee::AddSales(int salesCount)
{
	if (salesCount > 0)
	{
		m_salesCount += salesCount;
	}
}

bool Employee::OpenShift(const std::string& login, const std::string& password)
{
	// Проверка логина и пароля для открытия смены
	if (login == m_login && password == m_password)
	{
		std::cout << "Смена открыта для сотрудника: " << m_fullName << std::endl;
		return true;
	}
	else
	{
		std::cout << "Неверный логин или пароль!" << std::endl;
		return false;
	}
}

bool Employee::ChangeCredentials(const std::string& oldLogin, const std::string& oldPassword,
								 const std::string& newLogin, const std::string& newPassword)
{
	// Проверка старых учетных данных перед изменением
	if (oldLogin == m_login && oldPassword == m_password)
	{
		m_login = newLogin;
		m_password = newPassword;
		std::cout << "Учетные данные успешно изменены." << std::endl;
		return true;
	}
	else
	{
		std::cout << "Неверный текущий логин или пароль!" << std::endl;
		return false;
	}
}

std::string Employee::GetFullName() const
{
	return m_fullName;
}

int Employee::GetAge() const
{
	return m_age;
}

std::string Employee::GetContactNumber() const
{
	return m_contactNumber;
}

std::string Employee::GetAddress() const
{
	return m_address;
}

std::string Employee::GetPosition() const
{
	return m_position;
}

double Employee::GetHoursWorked() const
{
	return m_hoursWorked;
}

double Employee::GetHourlyRate() const
{
	return m_hourlyRate;
}

std::string Employee::GetLogin() const
{
	return m_login;
}

double Employee::GetSalaryBalance() const
{
	return m_salaryBalance;
}

int Employee::GetSalesCount() const
{
	return m_salesCount;
}

void Employee::SetFullName(const std::string& fullName)
{
	m_fullName = fullName;
}

void Employee::SetAge(int age)
{
	if (age >= 0)
	{
		m_age = age;
	}
}

void Employee::SetContactNumber(const std::string& contactNumber)
{
	m_contactNumber = contactNumber;
}

void Employee::SetAddress(const std::string& address)
{
	m_address = address;
}

void Employee::SetPosition(const std::string& position)
{
	m_position = position;
}

void Employee::SetHoursWorked(double hoursWorked)
{
	if (hoursWorked >= 0.0)
	{
		m_hoursWorked = hoursWorked;
	}
}

void Employee::SetHourlyRate(double hourlyRate)
{
	if (hourlyRate >= 0.0)
	{
		m_hourlyRate = hourlyRate;
	}
}

void Employee::SetSalaryBalance(double salaryBalance)
{
	m_salaryBalance = salaryBalance;
}

File: lab4/Employee.java
public class Employee {
    // Статическое поле: общее количество сотрудников
    private static int totalEmployees = 0;

    // Приватные поля (недоступны наследникам)
    private String login;             // Логин
    private String password;          // Пароль
    
    // Protected поля (доступны наследникам для демонстрации protected)
    protected String fullName;          // ФИО
    protected int age;                  // Возраст
    protected String contactNumber;     // Контактный номер
    protected String address;           // Адрес
    protected String position;          // Должность
    protected double hoursWorked;       // Отработанные часы
    protected double hourlyRate;        // Почасовая ставка
    protected double salaryBalance;     // Баланс з/п (для штрафов, премий, списаний)
    protected int salesCount;           // Количество продаж

    // Конструктор по умолчанию
    public Employee() {
        this.fullName = "";
        this.age = 0;
        this.contactNumber = "";
        this.address = "";
        this.position = "";
        this.hoursWorked = 0.0;
        this.hourlyRate = 0.0;
        this.login = "";
        this.password = "";
        this.salaryBalance = 0.0;
        this.salesCount = 0;
        totalEmployees++; // Увеличиваем счетчик сотрудников
    }

    // Конструктор с параметрами
    public Employee(String fullName, int age, String contactNumber,
                   String address, String position,
                   double hourlyRate, String login, String password) {
        this.fullName = fullName;
        this.age = age;
        this.contactNumber = contactNumber;
        this.address = address;
        this.position = position;
        this.hoursWorked = 0.0;
        this.hourlyRate = hourlyRate;
        this.login = login;
        this.password = password;
        this.salaryBalance = 0.0;
        this.salesCount = 0;
        totalEmployees++; // Увеличиваем счетчик сотрудников
    }

    // "Конструктор копирования" в Java (через другой конструктор)
    public Employee(Employee other) {
        this(other.fullName, other.age, other.contactNumber,
             other.address, other.position,
             other.hourlyRate, other.login, other.password);
        this.hoursWorked = other.hoursWorked;
        this.salaryBalance = other.salaryBalance;
        this.salesCount = other.salesCount;
    }

    // Расчет заработной платы
    public double calculateSalary() {
        // Расчет з/п = отработанные часы * почасовая ставка + баланс (премии/штрафы)
        return hoursWorked * hourlyRate + salaryBalance;
    }

    // Добавить продажи
    public void addSales(int salesCount) {
        if (salesCount > 0) {
            this.salesCount += salesCount;
        }
    }

    // Открыть смену (проверка логина/пароля)
    public boolean openShift(String login, String password) {
        // Проверка логина и пароля для открытия смены
        if (login.equals(this.login) && password.equals(this.password)) {
            System.out.println("Смена открыта для сотрудника: " + fullName);
            return true;
        } else {
            System.out.println("Неверный логин или пароль!");
            return false;
        }
    }

    // Сменить логин/пароль
    public boolean changeCredentials(String oldLogin, String oldPassword,
                                    String newLogin, String newPassword) {
        // Проверка старых учетных данных перед изменением
        if (oldLogin.equals(this.login) && oldPassword.equals(this.password)) {
            this.login = newLogin;
            this.password = newPassword;
            System.out.println("Учетные данные успешно изменены.");
            return true;
        } else {
            System.out.println("Неверный текущий логин или пароль!");
            return false;
        }
    }

    // Геттеры
    public String getFullName() {
        return fullName;
    }

    public int getAge() {
        return age;
    }

    public String getContactNumber() {
        return contactNumber;
    }

    public String getAddress() {
        return address;
    }

    public String getPosition() {
        return position;
    }

    public double getHoursWorked() {
        return hoursWorked;
    }

    public double getHourlyRate() {
        return hourlyRate;
    }

    public String getLogin() {
        return login;
    }

    public double getSalaryBalance() {
        return salaryBalance;
    }

    public int getSalesCount() {
        return salesCount;
    }

    // Сеттеры
    public void setFullName(String fullName) {
        this.fullName = fullName;
    }

    public void setAge(int age) {
        if (age >= 0) {
            this.age = age;
        }
    }

    public void setContactNumber(String contactNumber) {
        this.contactNumber = contactNumber;
    }

    public void setAddress(String address) {
        this.address = address;
    }

    public void setPosition(String position) {
        this.position = position;
    }

    public void setHoursWorked(double hoursWorked) {
        if (hoursWorked >= 0.0) {
            this.hoursWorked = hoursWorked;
        }
    }

    public void setHourlyRate(double hourlyRate) {
        if (hourlyRate >= 0.0) {
            this.hourlyRate = hourlyRate;
        }
    }

    public void setSalaryBalance(double salaryBalance) {
        this.salaryBalance = salaryBalance;
    }

    // Статический метод: получить общее количество сотрудников
    public static int getTotalEmployees() {
        return totalEmployees;
    }
}

